include/drivers/ipm.h

```diff
@@ -69,7 +105,8 @@ typedef void (*ipm_callback_t)(const struct device *ipmdev, void *user_data,
  *
  * See @a ipm_send() for argument definitions.
  */
-typedef int (*ipm_send_t)(const struct device *ipmdev, int wait, struct ipm_msg *msg);
+typedef int (*ipm_send_t)(const struct device *ipmdev, int wait, uint32_t channel,
+			  struct ipm_msg *msg);
```

Comment:
How do you ensure that the channel is the one declared in the DT for the driver usage?
Based on this API any driver could send a message on any channel of the hardware devide without control, right?

It seems to me that using a new handle would be safer. It could be a constant structure containing the ipmdev and the channel declared in the device tree, something like 
struct mbox_chan {
	const struct device *ipmdev;
	uint32_t channel;
	ipm_callback_t cb;
	void *user_data;
};
typedef int (*ipm_send_t)(const struct mbox_chan channel,  struct ipm_msg *msg); 
typedef void (*ipm_register_callback_t)(const struct mbox_chan channel,
					ipm_callback_t cb,
					void *user_data);
...

Reply:
> How do you ensure that the channel is the one declared in the DT for the driver usage?

The driver using the IPM API should retrieve the proper channel from the DT using the `DT_*` helpers on the `mbox` bindings. I can make that more explicit if needed.

> Based on this API any driver could send a message on any channel of the hardware devide without control, right?

Yes, but do we care? The user is in charge of the correctness of the DT for its own hardware from where the channel numbers used in this API are retrieved from. I don't think we should babysit the developers :)

> It seems to me that using a new handle would be safer.

I think this is an unnecessary complication TBH.

Reply:
> > * `#ipm-cells = <0>`: one instance of the mailbox per channel ( seems already exist in Zephyr)
> > * `#ipm-cells > <1>`. For example, a second parameter can be used for bidirectional mailboxes to specify RX or TX.
> 
> Just for clarity: `#ipm-cells` should really be `#mbox-cells`, the commit message is wrong, I'll fix that.
> 
> The difference between `#mbox-cells = <0>` and `#mbox-cells = <1>` is that in the first case the IPM driver will be just discarding whatever is passed as `channel` parameter, in the second case the driver will take that into accont.
> 
> > In this case, how to handle a generic client code to be able to support different IPM backends? Using a struct handle would make APi more flexible according to the ` #ipm-cells`.
> 
> You mean when an IPM user device must use two different IPM devices at the same time with different `#mbox-cells` sizes? Is this a common case?

My point here is a use case with `#mbox-cells = <2>` 
In https://github.com/carlocaione/zephyr/blob/upstream_ipm/samples/subsys/ipc/openamp/remote/src/main.c, how to address a mailbox defined with ` #mbox-cells = <2>`?

```
 ipm0: ipm@2a000 {
     compatible = "vnd,ipm0";
     #mbox-cells = <2>;
     ...
 };

 ipm-user@0 {
     compatible = "some,ipm-user";
     mboxes = <&ipm0 1 0>, <&ipm0 1 1>;  /* second parameter defines the RX or TX  */
     mbox-names=  "tx", "rx";
 };
```
I can see such definitions in Linux kernel DT. It can never happen in Zephyr but if happens, this would probably break the API. Or does this be addressed using the ipm_configure_channel? In such case can code be platform agnostic?

> 
> Without considering for now the DT part, if `IPM0` is the device using `#ipm-cells = <0>` and `IPM1` the device using `#ipm-cells = <1>` the user must do something like:
> 
> ```
> ipm_send(IPM0, <whatever_channel_it_is_discarded_anyway>)
> ipm_send(IPM1, <tx_channel_retrieved_from_dt>)
> ```
Right  if #mbox-cells cannot be greater than 1 your API is fine.

> 
> The real problem here is how do you describe this situation in the DT. How do you use `mboxes` and `mbox-names` to reference two devices with different `#mbox-cells`?
> 
> ```
> ipm0: ipm@2a000 {
>     compatible = "vnd,ipm0";
>     #mbox-cells = <0>;
>     ...
> };
> 
> ipm1: ipm@2b000 {
>     compatible = "vnd,ipm1";
>     #mbox-cells = <1>;
>     ...
> };
> 
> ipm-user@0 {
>     compatible = "some,ipm-user";
>     mboxes = <&ipm1 9>, <&ipm1 8>;
>     mbox-names=  "tx", "rx";
>     ...
>    
>     what about ipm0 ???
> };
> ```
I don't know if this usecase exists.

Reply:
> My point here is a use case with `#mbox-cells = <2>` In https://github.com/carlocaione/zephyr/blob/upstream_ipm/samples/subsys/ipc/openamp/remote/src/main.c, how to address a mailbox defined with ` #mbox-cells = <2>`?

Oooh ok, probably I got now your point (hopefully :)

As long as the IPM device has one of the cells named `channel` in the binding / YAML file (see https://github.com/carlocaione/zephyr/blob/upstream_ipm/dts/bindings/test/vnd%2Cipm.yaml#L15) then the DT helper (see https://github.com/carlocaione/zephyr/blob/upstream_ipm/include/drivers/ipm.h#L233) is able to retrieve the correct TX / RX channel number.

So, back to your example:
> ```
>  ipm0: ipm@2a000 {
>      compatible = "vnd,ipm0";
>      #mbox-cells = <2>;
>      ...
>  };
> 
>  ipm-user@0 {
>      compatible = "some,ipm-user";
>      mboxes = <&ipm0 1 0>, <&ipm0 1 1>;  /* second parameter defines the RX or TX  */
>      mbox-names=  "tx", "rx";
>  };
> ```

In this case I expect the binding for `vnd,ipm0` being something like:
```
...
mbox-cells:
  - whatever_parameter
  - channel
```
So in your code you can still do:
```
unsigned int tx_channel = DT_IPMS_CHANNEL_BY_NAME(DT_NODELABEL(ipm0), tx); // this will be 0
unsigned int rx_channel = DT_IPMS_CHANNEL_BY_NAME(DT_NODELABEL(ipm0), rx); // this will be 1
```
And use `tx_channel` and `rx_channel` with the API.

Does this make sense?

I will anyway extend the API to better support `ipm_dt_spec` so that we can retrieve channel and device using only one macro and use that directly as you suggested.